File: javaexamples/src/pc/ProducerConsumerSemaphore.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.ReentrantLock;

class SemaphoreBuffer {
    private final Queue<Integer> buffer;
    private final int capacity;
    final Semaphore empty;    // Tracks empty slots (producers can produce)
    final Semaphore full;     // Tracks full slots (consumers can consume)
    private final Semaphore mutex;    // Mutual exclusion for buffer access

    public SemaphoreBuffer(int capacity) {
        this.capacity = capacity;
        this.buffer = new LinkedList<>();
        this.empty = new Semaphore(capacity);  // Initially all slots are empty
        this.full = new Semaphore(0);          // Initially no items to consume
        this.mutex = new Semaphore(1);         // Binary semaphore for mutual exclusion
    }

    public void produce(int item) throws InterruptedException {
        empty.acquire();    // Wait for empty slot
        mutex.acquire();    // Enter critical section

        try {
            buffer.offer(item);
            System.out.println(Thread.currentThread().getName() + " produced: " + item +
                             " [Buffer size: " + buffer.size() + "]");
        } finally {
            mutex.release();    // Exit critical section
        }

        full.release();     // Signal that an item is available
    }

    public int consume() throws InterruptedException {
        full.acquire();     // Wait for available item
        mutex.acquire();    // Enter critical section

        int item;
        try {
            item = buffer.poll();
            System.out.println(Thread.currentThread().getName() + " consumed: " + item +
                             " [Buffer size: " + buffer.size() + "]");
        } finally {
            mutex.release();    // Exit critical section
        }

        empty.release();    // Signal that a slot is available
        return item;
    }

    public int size() throws InterruptedException {
        mutex.acquire();
        try {
            return buffer.size();
        } finally {
            mutex.release();
        }
    }

    public void printSemaphoreState() {
        System.out.println("Semaphore state - Empty permits: " + empty.availablePermits() +
                         ", Full permits: " + full.availablePermits() +
                         ", Mutex permits: " + mutex.availablePermits());
    }
}

class SemaphoreProducer implements Runnable {
    private final SemaphoreBuffer buffer;
    private final int itemCount;
    private final int producerId;

    public SemaphoreProducer(SemaphoreBuffer buffer, int itemCount, int producerId) {
        this.buffer = buffer;
        this.itemCount = itemCount;
        this.producerId = producerId;
    }

    @Override
    public void run() {
        try {
            for (int i = 1; i <= itemCount; i++) {
                int item = producerId * 100 + i;

                // Check if buffer might be full
                if (buffer.empty.availablePermits() == 0) {
                    System.out.println(Thread.currentThread().getName() + " waiting - buffer full");
                }

                buffer.produce(item);
                Thread.sleep(100 + (int)(Math.random() * 200)); // Random delay
            }
            System.out.println(Thread.currentThread().getName() + " finished producing");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.out.println(Thread.currentThread().getName() + " interrupted");
        }
    }
}

class SemaphoreConsumer implements Runnable {
    private final SemaphoreBuffer buffer;
    private final int itemCount;

    public SemaphoreConsumer(SemaphoreBuffer buffer, int itemCount) {
        this.buffer = buffer;
        this.itemCount = itemCount;
    }

    @Override
    public void run() {
        try {
            for (int i = 0; i < itemCount; i++) {
                // Check if buffer might be empty
                if (buffer.full.availablePermits() == 0) {
                    System.out.println(Thread.currentThread().getName() + " waiting - buffer empty");
                }

                int item = buffer.consume();
                Thread.sleep(150 + (int)(Math.random() * 250)); // Random delay
            }
            System.out.println(Thread.currentThread().getName() + " finished consuming");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.out.println(Thread.currentThread().getName() + " interrupted");
        }
    }
}

public class ProducerConsumerSemaphore {
    public static void main(String[] args) throws InterruptedException {
        final int BUFFER_SIZE = 5;
        final int ITEMS_PER_PRODUCER = 3;
        final int ITEMS_PER_CONSUMER = 2;
        final int NUM_PRODUCERS = 3;
        final int NUM_CONSUMERS = 4;

        SemaphoreBuffer buffer = new SemaphoreBuffer(BUFFER_SIZE);

        Thread[] producers = new Thread[NUM_PRODUCERS];
        Thread[] consumers = new Thread[NUM_CONSUMERS];

        System.out.println("Starting Producer-Consumer Demo with Semaphores");
        System.out.println("Buffer capacity: " + BUFFER_SIZE);
        System.out.println("Producers: " + NUM_PRODUCERS + " (each producing " + ITEMS_PER_PRODUCER + " items)");
        System.out.println("Consumers: " + NUM_CONSUMERS + " (each consuming " + ITEMS_PER_CONSUMER + " items)");
        System.out.println("Total items to produce: " + (NUM_PRODUCERS * ITEMS_PER_PRODUCER));
        System.out.println("Total items to consume: " + (NUM_CONSUMERS * ITEMS_PER_CONSUMER));
        System.out.println();

        buffer.printSemaphoreState();
        System.out.println();

        // Create producer threads
        for (int i = 0; i < NUM_PRODUCERS; i++) {
            producers[i] = new Thread(new SemaphoreProducer(buffer, ITEMS_PER_PRODUCER, i + 1));
            producers[i].setName("Producer-" + (i + 1));
        }

        // Create consumer threads
        for (int i = 0; i < NUM_CONSUMERS; i++) {
            consumers[i] = new Thread(new SemaphoreConsumer(buffer, ITEMS_PER_CONSUMER));
            consumers[i].setName("Consumer-" + (i + 1));
        }

        // Start all threads
        for (Thread producer : producers) {
            producer.start();
        }

        for (Thread consumer : consumers) {
            consumer.start();
        }

        // Wait for all threads to complete
        for (Thread producer : producers) {
            producer.join();
        }

        for (Thread consumer : consumers) {
            consumer.join();
        }

        System.out.println("\nDemo completed.");
        System.out.println("Final buffer size: " + buffer.size());
        buffer.printSemaphoreState();
    }
}

File: goexamples/concurrency/pc/producer_consumer_semaphore.go
package main

import (
	"context"
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Semaphore implementation using channels
type Semaphore struct {
	ch chan struct{}
}

func NewSemaphore(n int) *Semaphore {
	return &Semaphore{
		ch: make(chan struct{}, n),
	}
}

func (s *Semaphore) Acquire() {
	s.ch <- struct{}{}
}

func (s *Semaphore) Release() {
	select {
	case <-s.ch:
	default:
		// Channel is empty, nothing to release
	}
}

func (s *Semaphore) AvailablePermits() int {
	return cap(s.ch) - len(s.ch)
}

// SemaphoreBuffer represents a bounded buffer using semaphores
type SemaphoreBuffer struct {
	buffer []int
	mutex  *Semaphore // Binary semaphore for mutual exclusion
	empty  *Semaphore // Tracks empty slots
	full   *Semaphore // Tracks full slots
}

func NewSemaphoreBuffer(capacity int) *SemaphoreBuffer {
	return &SemaphoreBuffer{
		buffer: make([]int, 0, capacity),
		mutex:  NewSemaphore(1),           // Binary semaphore
		empty:  NewSemaphore(capacity),    // Initially all slots empty
		full:   NewSemaphore(0),           // Initially no items
	}
}

func (sb *SemaphoreBuffer) Produce(item int, producerID int) {
	sb.empty.Acquire() // Wait for empty slot
	sb.mutex.Acquire() // Enter critical section

	sb.buffer = append(sb.buffer, item)
	fmt.Printf("Producer-%d produced: %d [Buffer size: %d]\n", producerID, item, len(sb.buffer))

	sb.mutex.Release() // Exit critical section
	sb.full.Release()  // Signal item available
}

func (sb *SemaphoreBuffer) Consume(consumerID int) int {
	sb.full.Acquire()  // Wait for available item
	sb.mutex.Acquire() // Enter critical section

	item := sb.buffer[0]
	sb.buffer = sb.buffer[1:]
	fmt.Printf("Consumer-%d consumed: %d [Buffer size: %d]\n", consumerID, item, len(sb.buffer))

	sb.mutex.Release() // Exit critical section
	sb.empty.Release() // Signal slot available
	return item
}

func (sb *SemaphoreBuffer) Size() int {
	sb.mutex.Acquire()
	defer sb.mutex.Release()
	return len(sb.buffer)
}

func (sb *SemaphoreBuffer) PrintSemaphoreState() {
	fmt.Printf("Semaphore state - Empty permits: %d, Full permits: %d, Mutex permits: %d\n",
		sb.empty.AvailablePermits(), sb.full.AvailablePermits(), sb.mutex.AvailablePermits())
}

func producer(ctx context.Context, buffer *SemaphoreBuffer, producerID, itemCount int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := 1; i <= itemCount; i++ {
		select {
		case <-ctx.Done():
			fmt.Printf("Producer-%d interrupted\n", producerID)
			return
		default:
			item := producerID*100 + i

			// Check if buffer might be full
			if buffer.empty.AvailablePermits() == 0 {
				fmt.Printf("Producer-%d waiting - buffer full\n", producerID)
			}

			buffer.Produce(item, producerID)

			// Random delay
			time.Sleep(time.Duration(100+rand.Intn(200)) * time.Millisecond)
		}
	}
	fmt.Printf("Producer-%d finished producing\n", producerID)
}

func consumer(ctx context.Context, buffer *SemaphoreBuffer, consumerID, itemCount int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := 0; i < itemCount; i++ {
		select {
		case <-ctx.Done():
			fmt.Printf("Consumer-%d interrupted\n", consumerID)
			return
		default:
			// Check if buffer might be empty
			if buffer.full.AvailablePermits() == 0 {
				fmt.Printf("Consumer-%d waiting - buffer empty\n", consumerID)
			}

			buffer.Consume(consumerID)

			// Random delay
			time.Sleep(time.Duration(150+rand.Intn(250)) * time.Millisecond)
		}
	}
	fmt.Printf("Consumer-%d finished consuming\n", consumerID)
}

func main() {
	const (
		bufferSize       = 5
		itemsPerProducer = 3
		itemsPerConsumer = 2
		numProducers     = 3
		numConsumers     = 4
	)

	buffer := NewSemaphoreBuffer(bufferSize)
	var wg sync.WaitGroup

	fmt.Println("Starting Producer-Consumer Demo with Semaphores (Go)")
	fmt.Printf("Buffer capacity: %d\n", bufferSize)
	fmt.Printf("Producers: %d (each producing %d items)\n", numProducers, itemsPerProducer)
	fmt.Printf("Consumers: %d (each consuming %d items)\n", numConsumers, itemsPerConsumer)
	fmt.Printf("Total items to produce: %d\n", numProducers*itemsPerProducer)
	fmt.Printf("Total items to consume: %d\n", numConsumers*itemsPerConsumer)
	fmt.Println()

	buffer.PrintSemaphoreState()
	fmt.Println()

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// Start producers
	for i := 1; i <= numProducers; i++ {
		wg.Add(1)
		go producer(ctx, buffer, i, itemsPerProducer, &wg)
	}

	// Start consumers
	for i := 1; i <= numConsumers; i++ {
		wg.Add(1)
		go consumer(ctx, buffer, i, itemsPerConsumer, &wg)
	}

	// Wait for all goroutines to complete
	wg.Wait()

	fmt.Println("\nDemo completed.")
	fmt.Printf("Final buffer size: %d\n", buffer.Size())
	buffer.PrintSemaphoreState()
}